### ISSUE TITLE
Automatic disc resize

### ISSUE THREAD

**Issue by rostwald** [morganwdavis/freebsd-digitalocean#5]
Discussion about improved logic for disc resize.

_Originally posted by @morganwdavis in https://github.com/morganwdavis/freebsd-digitalocean/issues/3#issuecomment-486145800_

> 
> ```
> # gpart show -l
> =>      34  83886013  vtbd0  GPT  (40G)
>         34         6         - free -  (3.0K)
>         40      1024      1  gptboot0  (512K)
>       1064       984         - free -  (492K)
>       2048   4194304      2  swap0  (2.0G)
>    4196352  79687680      3  zfs0  (38G)
>   83884032      2015         - free -  (1.0M)
> ```
> ```
> # zpool status
>   pool: zroot
>  state: ONLINE
>   scan: scrub repaired 0 in 0 days 00:00:42 with 0 errors on Wed Apr 10 22:49:33 2019
> config:
> 
>         NAME        STATE     READ WRITE CKSUM
>         zroot       ONLINE       0     0     0
>           vtbd0p3   ONLINE       0     0     0
> 
> errors: No known data errors
> ```
> ```
> # ls -l /dev/gpt
> total 0
> crw-r-----  1 root  operator  0x35 Apr  1 22:16 gptboot0
> ```
> And here are the commands I have to issue to expand a resized droplet:
> ```
> # gpart recover vtbd0
> # gpart resize -i 3 vtbd0
> # zpool set autoexpand=on zroot
> # zpool online -e zroot vtbd0p3
> ```

I think combining some checks via gpart and zpool status might be the way to go. E.g. probing (all?) devices via `gpart show` and finding the zfs partitions/slices; checking for additional "free" space and finally probing `zpool status` whether the found partitions that could be extended are actually used in a pool (or allowing per-pool rules in the configuration).
This way we can also catch pools on attached block storage volumes (I use such a configuration e.g. on mail servers).

Should we also support automatic resizing for UFS droplets? Usually the partition layout with UFS is more complex and we can't safely assume which partition(s) to extend...

**Comment by morganwdavis** [comment]
Yes, parsing `gpart show` output should give hints to identify the device/geom name and partition indexes. Example using a droplet that has both local ZFS storage and block storage volume:

```
# gpart show
=>      34  83886013  vtbd0  GPT  (40G)
        34         6         - free -  (3.0K)
        40      1024      1  freebsd-boot  (512K)
      1064       984         - free -  (492K)
      2048   4194304      2  freebsd-swap  (2.0G)
   4196352  79687680      3  freebsd-zfs  (38G)
  83884032      2015         - free -  (1.0M)

=>       34  629145526  da0  GPT  (300G)
         34  629145526    1  freebsd-zfs  (300G)
```

We could also use the presence of the '[CORRUPT]' flag to know if any work even needs to be done. (Per the handbook example: https://www.freebsd.org/doc/handbook/disks-growing.html)

From that, we can then know to do the `gpart recover <geom>`.

For the partitions, we would need to do a `gpart show <geom>` and then walk the output to find the index associated with the freebsd-zfs entry. If we find more than one, I don't think we can do the autoexpand since we can't know how the admin prefers to distribute the extra space. So single freebsd-zfs partitions would qualify and from that we'd be able to do the `gpart resize -i <index> <geom>`.

What I'm not sure about is mapping a gpart device and partition to its associated ZFS pool in order to do the `zpool online <pool> <device>` command. Parsing `zpool status` to match the device to a pool seems messy. Is there a better/cleaner way?

For me, UFS is an afterthought as I've stopped using it. Someone else will need to take that on. :)
